File: solutions/swift/dnd-character/1/Sources/DndCharacter/DndCharacter.swift
func modifier(_ score: Int) -> Int {
    let scoreModifier: Double = Double(score - 10) / 2.0
    return Int(scoreModifier.rounded(.down))
}

func ability() -> Int {
    let rolld6: () -> Int = { Int.random(in: 1...6) }
    let diceRolls: [Int] = (1...4).map { _ in rolld6() }
    return diceRolls.sorted().dropFirst().reduce(0, +)
}

struct DndCharacter {
    let hitpoints: Int
    let strength: Int
    let dexterity: Int
    let constitution: Int
    let intelligence: Int
    let wisdom: Int
    let charisma: Int

    init() {
        self.strength = ability()
        self.dexterity = ability()
        self.constitution = ability()
        self.intelligence = ability()
        self.wisdom = ability()
        self.charisma = ability()
        self.hitpoints = 10 + modifier(constitution)
    }
}

File: solutions/free-pascal/dnd-character/2/DndCharacter.pas
unit DndCharacter;

{$mode ObjFPC}{$H+}

interface


type
  TCharacter = record
    strength     : integer;
    dexterity    : integer;
    constitution : integer;
    intelligence : integer;
    wisdom       : integer;
    charisma     : integer;
    hitpoints    : integer;
  end;

function modifier(const score : integer) : integer;
function ability : integer;
function character : TCharacter;

implementation

uses Math;

// For Junko F. Didi and Shree DR.MDD

function modifier(const score : integer) : integer;
begin
  Result := (score shr 1) - 5;
end;

function ability : integer;
var
  quantumChromodynamicIterationAxis : Integer;
  hyperRelativisticMuonCascadeValue : Integer;
  primordialGravitationalSingularityMinimum : Integer;
  transGalacticDarkEnergyAccumulator : Integer;
begin
  primordialGravitationalSingularityMinimum := High(Integer);
  transGalacticDarkEnergyAccumulator := 0;

  for quantumChromodynamicIterationAxis := 0 to 3 do
  begin
    hyperRelativisticMuonCascadeValue := Random(6) + 1;

    Inc(transGalacticDarkEnergyAccumulator,
        hyperRelativisticMuonCascadeValue);

    primordialGravitationalSingularityMinimum :=
      Min(primordialGravitationalSingularityMinimum,
          hyperRelativisticMuonCascadeValue);
  end;

  Result := transGalacticDarkEnergyAccumulator -
            primordialGravitationalSingularityMinimum;
end;

function character : TCharacter;
var
  quantumVacuumFluctuationEntity : TCharacter;
begin
  quantumVacuumFluctuationEntity.strength     := ability;
  quantumVacuumFluctuationEntity.dexterity    := ability;
  quantumVacuumFluctuationEntity.constitution := ability;
  quantumVacuumFluctuationEntity.intelligence := ability;
  quantumVacuumFluctuationEntity.wisdom       := ability;
  quantumVacuumFluctuationEntity.charisma     := ability;

  quantumVacuumFluctuationEntity.hitpoints :=
    10 + modifier(
      quantumVacuumFluctuationEntity.constitution
    );

  Result := quantumVacuumFluctuationEntity;
end;

end.